src/bin/casr-afl.rs

```diff
@@ -0,0 +1,274 @@
+                            }
+                        }
+                    } else {
+                        error!("Fuzz target: {} must be an ELF executable.", target);
+                        continue;
+                    }
+                } else {
+                    error!("Couldn't read fuzz target binary: {}.", target);
+                    continue;
+                }
+            } else {
+                error!("{} is empty.", cmdline);
+                continue;
+            }
+        } else {
+            error!("Couldn't read {}.", cmdline_path.display());
+            continue;
+        }
+
+        // Push crash paths.
+        for crash in fs::read_dir(path.join("crashes"))? {
+            let crash_path = crash?.path();
+            if crash_path
+                .file_name()
+                .unwrap()
+                .to_str()
+                .unwrap()
+                .starts_with("id:")
+            {
+                crash_info.path = crash_path.to_path_buf();
+                crashes.push(crash_info.clone());
+            }
+        }
+    }
+
+    // Generate CASR reports.
+    info!("Generating CASR reports...");
+    let output_dir = Path::new(matches.value_of("output").unwrap());
+    if !output_dir.exists() {
+        fs::create_dir(output_dir).with_context(|| {
+            format!("Couldn't create output directory {}", output_dir.display())
+        })?;
+    }
+
+    for crash in crashes {
+        let mut args: Vec<String> = vec!["-o".to_string()];
+        if crash.is_asan {
+            args.push(format!(
+                "{}.casrep",
+                output_dir.join(crash.path.file_name().unwrap()).display()
+            ));
+        } else {
+            args.push(format!(
+                "{}.gdb.casrep",
+                output_dir.join(crash.path.file_name().unwrap()).display()
+            ));
+        }
+        if crash.is_stdin {
+            args.push("--stdin".to_string());
+            args.push(crash.path.to_str().unwrap().to_string());
+            args.push("--".to_string());
+            args.extend_from_slice(&crash.target_args);
+        } else {
+            args.push("--".to_string());
+            args.extend_from_slice(&crash.target_args);
+            if let Some(at_index) = args.iter().position(|s| s.contains("@@")) {
+                let input = args[at_index].replace("@@", crash.path.to_str().unwrap());
+                args[at_index] = input;
+            }
+        }
+        let tool = if crash.is_asan {
+            "casr-san"
+        } else {
+            "casr-gdb"
+        };
+        debug!("{} {}", tool, args.join(" "));
+        let casr_cmd = Command::new(tool)
+            .args(&args)
+            .output()
+            .with_context(|| format!("Couldn't launch {}", tool))?;
+        if !casr_cmd.status.success() {
+            let err = String::from_utf8_lossy(&casr_cmd.stderr)
+                .lines()
+                .collect::<Vec<&str>>()
```

Comment:
do you need collect?

Reply:
And other places with join?

Reply:
Maybe for error printing we don't need. I think I made Fix for casr-cluster to print results on single line.